piet-cairo/src/text.rs

```diff
@@ -210,233 +396,198 @@ impl TextLayout for CairoTextLayout {
+            }
+
+            let line_text = &self.text[start_offset..end_offset];
+            let trimmed_len = line_text.trim_end().len();
+            let trailing_whitespace = line_text[trimmed_len..].len();
+
+            widest_whitespaceless_width = {
+                let start_x = line.index_to_x(trimmed_len as i32, false);
+                let end_x = line.index_to_x(line_text.len() as i32, true);
+                let whitespace_width = end_x - start_x;
+                let whitespaceless_width = logical_rect.width - whitespace_width;
+                widest_whitespaceless_width.max(whitespaceless_width)
+            };
+
+            line_metrics.push(LineMetric {
+                start_offset,
+                end_offset,
+                trailing_whitespace,
+                baseline: (iterator.get_baseline() as f64 / PANGO_SCALE) - y_offset,
+                height: logical_rect.height as f64 / PANGO_SCALE,
+                y_offset,
+            });
+            y_offset += logical_rect.height as f64 / PANGO_SCALE;
+
+            if !iterator.next_line() {
+                break;
+            }
+        }
 
-    let text_len = text.len();
+        //NOTE: Pango appears to always give us at least one line even with empty input
+        self.line_metrics = line_metrics;
 
-    if text_position == 0 {
-        return 0.0;
-    }
+        let ink_extent = self.pango_layout.get_extents().0;
+        let logical_extent = self.pango_layout.get_extents().1;
+        self.size = Size::new(
+            widest_whitespaceless_width as f64 / PANGO_SCALE,
+            logical_extent.height as f64 / PANGO_SCALE,
+        );
+        self.ink_size = Size::new(
+            ink_extent.width as f64 / PANGO_SCALE,
+            ink_extent.height as f64 / PANGO_SCALE,
+        );
+        self.pango_offset = Vec2::new(
+            logical_extent.x as f64 / PANGO_SCALE,
+            logical_extent.y as f64 / PANGO_SCALE,
+        );
 
-    if text_position as usize >= text_len {
-        return font.text_extents(&text).x_advance;
+        self.trailing_ws_width = widest_logical_width as f64 / PANGO_SCALE;
     }
-
-    // Already checked that text_position > 0 and text_position < count.
-    // If text position is not at a grapheme boundary, use the text position of current
-    // grapheme cluster. But return the original text position
-    // Use the indices (byte offset, which for our purposes = utf8 code units).
-    let grapheme_indices = UnicodeSegmentation::grapheme_indices(text, true)
-        .take_while(|(byte_idx, _s)| text_position >= *byte_idx);
-
-    grapheme_indices
-        .last()
-        .map(|(idx, _)| font.text_extents(&text[..idx]).x_advance)
-        .unwrap_or_else(|| font.text_extents(&text).x_advance)
 }
 
-fn scale_matrix(scale: f64) -> Matrix {
-    Matrix {
-        xx: scale,
-        yx: 0.0,
-        xy: 0.0,
-        yy: scale,
-        x0: 0.0,
-        y0: 0.0,
+impl fmt::Debug for CairoTextLayout {
```

Comment:
If I'm going to impl this I might put a bit more info in it? line count, original text, I could even imagine printing each line's text and metrics... 🤷

Reply:
There are several such impls of `fmt::Debug` but they are all inherited from the existing code